### PR TITLE
Improvements for Server Profiles handling of Templates

### LINKED ISSUE #156: Server Profile does not fill default values for OS Custom Attributes automatically
### Scenario/Intent
This is related to enhancement [234](https://github.com/HewlettPackard/oneview-chef/issues/234) on our Chef repo. 

Reproducing description for this repo:
If some attribute that does have a default value is not passed in the osDeploymentSettings' osCustomAttributes, it appears as not set in OneView instead assuming its default value.

### Environment Details
 - **Puppet module for HPE OneView Version:** v2.2.1
 - **OneView Version:** API300

### Steps to Reproduce
1. Try to apply a server profile with a valid os_deployment_plan with custom attributes that have default values.
2. Do not define these default value attributes in the OS custom attributes inside 'data' => 'osDeploymentSettings' => 'osCustomAttributes'.

#### Expected Result
The attributes that weren't specified should appear with default values, if they have. 

#### Actual Result
The attributes that weren't specified appears as Not Set, even if they have default values.

### Root cause analysis
OneView does not sets these values, we need to perform a get then override only the ones specified by the user in the osCustomAttributes before using the SDK's set_os_deployment_settings.

### LINKED ISSUE #165: Creating a Server Profile based on a Template does not autofill its blank attributes with the template's
### Scenario/Intent
Create a new `Server Profile` by specifying only its `name` and the `serverProfileTemplateUri` to which it will be associated, but having all other information retrieved directly from the template.

### Environment Details
 - **Puppet module for HPE OneView Version:** 2.2.2

### Steps to Reproduce
Run a manifest with something similar to the following:

```puppet
oneview_server_profile{'Server Profile Create':
  ensure => 'present',
  data   =>
  {
    name => $server_profile_name,
    serverProfileTemplateUri => $server_profile_template_name,
    serverHardwareUri => $server_hardware_name,
  }
}
```

### Expected Result
It should get other required information, such as `enclosureGroupUri` or `serverHardwareTypeUri` from the template if those are specified there.

### Actual Result
It fails because these 'required' fields were not passed in.

### PR DESCRIPTION
**Pull request by fgbulsoni**
### Description
Improvements for Server Profiles handling of Templates
- SP resources which have a serverProfileTemplateUri declared now, will
inherit all undeclared attributes from its template
- Added load_resource internal method to improve retrieval of different
resources
- Added load_template method to common to be used by SP and SPTs at the
moment

### Issues Resolved
#165 

### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass (`$ rake test`).
- [ ] New functionality has been documented in the README if applicable.
  - [ ] New functionality has been thoroughly documented in the examples (please include helpful comments).
- [x] Changes are documented in the CHANGELOG.

### CODE REVIEW

File: lib/puppet/provider/common.rb [hunk 1]
```diff
@@ -128,3 +128,37 @@ def connections_parse
     conn['networkUri'] = net.first['uri']
   end
 end
+
+# Retrieve a resource by type and identifier (name or data)
+# @param type [String, Symbol] Type of resource to be retrieved. e.g., :GoldenImage, :FCNetwork
+# @param id [String, Symbol, Hash] Name of the resource or Hash of data to retrieve by.
+#   Examples: 'EthNet1', { uri: '/rest/fake/123ABC' }
+# @param ret_attribute [NilClass, String, Symbol] If specified, returns a specific attribute of the resource.
```
Comment (aalexmonteiro) [inline lib/puppet/provider/common.rb]: ```base_module``` parameter is missing in documentation`.

**Review by aalexmonteiro — APPROVED** [review]
It looks good, although I don't know the architecture well.

File: lib/puppet/provider/common.rb [hunk 2]
```diff
@@ -128,3 +128,37 @@ def connections_parse
...
+  data = id.is_a?(Hash) ? id : { name: id }
+  r = klass.new(@client, data)
+  raise(OneviewSDK::NotFound, "#{type} with data '#{data}' was not found") unless r.retrieve!
+  return r unless ret_attribute
+  r[ret_attribute]
+end
+
+# Retrieve the template to be used for the resource and autofills blank fields with template's values
+# @param uri [String] Uri of the resource
+# @param uri [String] Method specific to the resource which retrieves the require template
```
Comment (ricardogpsf) [inline lib/puppet/provider/common.rb]: `@param` ~~uri~~` method [String] Method specific to...`

**Review by ricardogpsf — COMMENTED** [review]
Just one simple adjust. After that, good for me.

**Review by aalexmonteiro — APPROVED** [review]

**Pull request merged**